gobblin-data-management/src/main/java/gobblin/data/management/retention/version/finder/DatasetVersionFinder.java

```diff
@@ -0,0 +1,84 @@
+
+import gobblin.data.management.retention.dataset.Dataset;
+import gobblin.data.management.retention.version.DatasetVersion;
+import gobblin.data.management.util.PathUtils;
+
+
+/**
+ * Class to find dataset versions in the file system.
+ *
+ * Concrete subclasses should implement a ({@link org.apache.hadoop.fs.FileSystem}, {@link azkaban.utils.Props})
+ * constructor to be instantiated by {@link gobblin.data.management.retention.DatasetCleaner}.
+ *
+ * @param <T> Type of {@link gobblin.data.management.retention.version.DatasetVersion} expected from this class.
+ */
+public abstract class DatasetVersionFinder<T extends DatasetVersion> implements VersionFinder<T> {
+
+  FileSystem fs;
+
+  public DatasetVersionFinder(FileSystem fs, Props props) {
+    this.fs = fs;
+  }
+
+  /**
+   * Find dataset versions in the input {@link org.apache.hadoop.fs.Path}. Dataset versions are subdirectories of the
+   * input {@link org.apache.hadoop.fs.Path} representing a single manageable unit in the dataset.
+   * See {@link gobblin.data.management.retention.DatasetCleaner} for more information.
+   *
+   * @param dataset {@link org.apache.hadoop.fs.Path} to directory containing all versions of a dataset.
+   * @return Map of {@link gobblin.data.management.retention.version.DatasetVersion} and {@link org.apache.hadoop.fs.FileStatus}
+   *        for each dataset version found.
+   * @throws IOException
+   */
+  @Override
+  public Collection<T> findDatasetVersions(Dataset dataset) throws IOException {
+    Path versionGlobStatus = new Path(dataset.datasetRoot(), globVersionPattern());
+
+    FileStatus[] dataSetVersionPaths = this.fs.globStatus(versionGlobStatus);
+
+    List<T> dataSetVersions = Lists.newArrayList();
+    for(FileStatus dataSetVersionPath: dataSetVersionPaths) {
+      T datasetVersion = getDatasetVersion(
+          PathUtils.relativizePath(dataSetVersionPath.getPath(), dataset.datasetRoot()), dataSetVersionPath.getPath());
+      if(datasetVersion != null) {
+        dataSetVersions.add(datasetVersion);
+      }
+    }
+
+    return dataSetVersions;
+  }
+
+  /**
+   * Should return class of T.
+   */
+  public abstract Class<? extends DatasetVersion> versionClass();
+
+  /**
+   * Glob pattern relative to the root of the dataset used to find {@link org.apache.hadoop.fs.FileStatus} for each
+   * dataset version.
+   * @return glob pattern relative to dataset root.
+   */
+  public abstract Path globVersionPattern();
+
+  /**
+   * Parse {@link gobblin.data.management.retention.version.DatasetVersion} from the path of a dataset version.
+   * @param pathRelativeToDatasetRoot {@link org.apache.hadoop.fs.Path} of dataset version relative to dataset root.
+   * @param fullPath full {@link org.apache.hadoop.fs.Path} of the dataset version.
+   * @return {@link gobblin.data.management.retention.version.DatasetVersion} for that path.
+   */
+  public abstract T getDatasetVersion(Path pathRelativeToDatasetRoot, Path fullPath);
```

Comment:
The default constraint here is that some comparable information needs to be contained in `pathRelativeToDatasetRoot`. This may not work well when we purely run retention based on modifiedTimeStamp (for example). I am fine with the current implementation. But it may be better to have a generic comparable object as the first parameter in this method, instead of `Path`.